crates/e2e-testing/src/spin.rs

```diff
@@ -65,6 +66,16 @@ pub fn build_app(appname: &str) -> Result<Output> {
     utils::run(vec!["spin", "build"], Some(&appdir), None)
 }
 
+pub fn state_dir(appname: &str) -> String {
```

Comment:
It seems like you've moved away from the approach of using a temporary directory.  Could you say something about why that didn't work out please?  It would be useful to understand.  Thanks!

Reply:
We cannot initialize `tempdir` in `spin_controller` as it will go out of scope (and hence will get cleaned up) soon after calling `spin up`. 

to avoid that, we will have to add it to the function where we are initializing the test case itself (in `mod.rs`). But because we are using same set of test cases for `spin deploy`, now those will fail as `spin deploy` does not support `--state-dir`.

one way could be to update `cloud controller` (the implementation of `trait Controller` to run tests against cloud) to ignore the state-dir flags (by iterating over the `deploy_args`), but that will add more complexity to it. 

I hope it makes sense. I am still very open to hear if there are other ideas that can help here.

Reply:
Would it be possible to make the `TempDir` object local to `TestCase::run`, rather than a member of the controller?  Then rather than passing it as a flag, you could pass it as an argument into `Controller::run_app`: the local controller would compose a `--state-dir` argument, and the cloud controller would ignore it.  And drop the `TempDir` only once `spin up` has returned and `TestCase::run` is exiting.

Disclaimer: insert the "I don't know what I'm doing" dog here.  You know this part of the code _waaay_ better than I do.

Reply:
yeah that sounds like a fair compromise to me. I have made the change for this. thank you for working with me on this @itowlson, it is really appreciated.